src/main/kotlin/org/misarch/returns/service/ReturnService.kt

```diff
@@ -0,0 +1,100 @@
+import java.time.Duration
+import java.time.OffsetDateTime
+
+/**
+ * Service for [ReturnEntity]s
+ *
+ * @param repository the provided repository
+ * @param orderItemRepository the order item repository
+ * @param shipmentRepository the shipment repository
+ * @param orderRepository the order repository
+ * @param productVariantVersionRepository the product variant version repository
+ * @param eventPublisher the event publisher
+ */
+@Service
+class ReturnService(
+    repository: ReturnRepository,
+    private val orderItemRepository: OrderItemRepository,
+    private val shipmentRepository: ShipmentRepository,
+    private val orderRepository: OrderRepository,
+    private val productVariantVersionRepository: ProductVariantVersionRepository,
+    private val eventPublisher: EventPublisher
+) : BaseService<ReturnEntity, ReturnRepository>(repository) {
+
+    /**
+     * Creates a return
+     * Handles permission checking
+     *
+     * @param returnInput the return to create
+     * @param authorizedUser the authorized user, used for permission checking
+     * @return the created return
+     */
+    suspend fun createReturn(returnInput: CreateReturnInput, authorizedUser: AuthorizedUser): ReturnEntity {
+        val orderItems = orderItemRepository.findAllById(returnInput.orderItemIds).collectList().awaitSingle()
+        validateReturn(returnInput, orderItems, authorizedUser)
+        val orderId = orderItems.first().orderId
+        val refundedAmount = orderItems.sumOf { it.compensatableAmount!! }
+        val returnEntity = ReturnEntity(
+            reason = returnInput.reason, orderId = orderId, refundedAmount = refundedAmount, id = null
+        )
+        val savedReturn = repository.save(returnEntity).awaitSingle()
+        for (orderItem in orderItems) {
+            orderItem.returnedWithId = savedReturn.id
+        }
+        orderItemRepository.saveAll(orderItems).collectList().awaitSingle()
+        eventPublisher.publishEvent(
+            ReturnEvents.RETURN_CREATED,
+            ReturnDTO(savedReturn.id!!, orderId, returnInput.orderItemIds, returnInput.reason, refundedAmount)
+        )
+        return savedReturn
+    }
+
+    /**
+     * Checks that a return can be created
+     * Checks that at least one order item is provided, and all order items exist, belong to the same order,
+     * have been delivered and have not been returned yet
+     *
+     * @param returnInput the return to validate
+     * @param orderItems the order items to validate
+     * @param authorizedUser the authorized user, used for permission checking
+     */
+    private suspend fun validateReturn(
+        returnInput: CreateReturnInput, orderItems: MutableList<OrderItemEntity>, authorizedUser: AuthorizedUser
+    ) {
+        val orderId = orderItems.first().orderId
+        val order = orderRepository.findById(orderId).awaitSingle()
+        require(order.userId == authorizedUser.id) { "Only the original buyer can create returns" }
```

Comment:
this one not really, as the id of the authorized user does not make its way to the db (we do not save the returning user as it would be redundant)
however some of the other can and will be done